### PR TITLE
Documentation change: Now showing the difference between `map` and `and_then` with an example.

### PR DESCRIPTION
**Pull request by sandeep-datta**
Updated documentation to clarify the difference between `and_then` and `map`. This also explains why we need `and_then` in addition to `map`. Please look at the diff for more information.

r?  @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
Thanks! Could you also wrap the lines to 80 chars (appears to be the surrounding style). For the actual content, though:

r? @steveklabnik

**Comment by SDX2000** [comment]
I just felt the manually inserted 80 char wide line breaks are an anachronism in this day and age. Almost all editors can do word wrapping now. The reader should be given a choice of where they want to see the line breaks instead of making this decision for them. IMO line breaks should only be used to delineate paragraphs.

Also I think unnecessary hard line breaks will make whitespace significant during file merge/comparison and make it unnecessarily hard to contribute. For example you might overflow the 80 char limit when adding a word to a line and end up with a cascading change which affects multiple lines thus making it difficult to see what was actually changed. There may be other complications arising out of this decision.

I vote for removing hard line breaks from the document altogether. But I do not want to upset the apple cart on the other hand so please let me know if you would like to go with ...

1. Removing hard line breaks from the whole document or...
2. Uniformly breaking lines at the 80 char limit (for my changes only).

**Comment by alexcrichton** [comment]
Er sorry but this is just the style we use, changing it isn't really within the scope of this PR.

**Comment by SDX2000** [comment]
@alexcrichton, changed to 80 char column now.

**Comment by steveklabnik** [comment]
@bors: r+ rollup

Thank you!

File: src/doc/book/error-handling.md [hunk 1]
```diff
@@ -351,11 +351,28 @@ fn file_name(file_path: &str) -> Option<&str> {
...
 
 You might think that we could use the `map` combinator to reduce the case
-analysis, but its type doesn't quite fit. Namely, `map` takes a function that
-does something only with the inner value. The result of that function is then
-*always* [rewrapped with `Some`](#code-option-map). Instead, we need something
-like `map`, but which allows the caller to return another `Option`. Its generic
-implementation is even simpler than `map`:
+analysis, but its type doesn't quite fit...
+
+```rust
```
Comment (steveklabnik) [inline src/doc/book/error-handling.md]: this needs to be marked `ignore` as well, so that it doesn't get tested.

**Comment by SDX2000** [comment]
@steveklabnik please review and commit.

**Comment by steveklabnik** [comment]
@bors: r+ rollup

thank you!

**Pull request merged**